### ISSUE TITLE
mysql: data correctness issue with stream of large dataset

### ISSUE THREAD

**Issue by nrainer-materialize** [MaterializeInc/materialize#25124]
### What version of Materialize are you using?

7a402c18b6

### What is the issue?

https://github.com/MaterializeInc/materialize/pull/25052 introduces the `MySqlStreaming` in the feature benchmark, which, however, reveals correctness issues.
Failure instances in nightly builds are [1](https://buildkite.com/materialize/nightlies/builds/6375#018d8d00-d977-45be-b7b4-c6c3aebf76cb), [2](https://buildkite.com/materialize/nightlies/builds/6375#018d8d5f-d606-4182-9a0f-f7576a697add), and [3](https://buildkite.com/materialize/nightlies/builds/6375#018d8dfc-c66c-48c1-a14d-6ad1bdcf4a3c). I also managed to reproduce it locally by running `bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --other-tag latest --scenario MySqlStreaming` on branch [tests/mysql-benchmark](https://github.com/nrainer-materialize/materialize/tree/tests/mysql-benchmark) of [PR#25052.](https://github.com/MaterializeInc/materialize/pull/25052)

The scenario inserts 100'000 rows in chunks of 100 rows into table `t1` in MySQL. Table `t1` is defined as `CREATE TABLE t1 (pk SERIAL PRIMARY KEY, f2 BIGINT);` [(more details in the code)](https://github.com/MaterializeInc/materialize/blob/e46663511f7bd3677746f19e77b43876ce30a5de/test/feature-benchmark/scenarios.py#L1487).

All 100'000 rows arrive in MySQL:
```
mysql> select count(*), count(distinct pk) from t1;
+----------+--------------------+
| count(*) | count(distinct pk) |
+----------+--------------------+
|   100000 |             100000 |
+----------+--------------------+
1 row in set (0.02 sec)
```

However, mz shows duplicates!

```
materialize=> select count(*) from t1;
 count  
--------
 105300
(1 row)
```

```
materialize=> select * from t1;
   pk   | f2  
--------+-----
 1      |   1
 1      |   1
 2      |   2
 2      |   2
 3      |   3
 3      |   3
[...]
```

It is interesting that mz does not correctly group them. Could this be due to the optimizer knowing that column `pk` is supposed to be unique (which it is in MySQL)?

```
materialize=> select pk from t1 group by pk having count(*) > 1;
 pk 
----
(0 rows)

materialize=> select * from t1 where pk = 141;
 pk  | f2 
-----+----
 141 | 14
 141 | 14
(2 rows)

materialize=> select pk, count(*) from t1 where pk = 141 group by pk;
 pk  | count 
-----+-------
 141 |     1
 141 |     1
(2 rows)
```

This does not apply to `count(distinct)`:
```
materialize=> select count(*), count(distinct pk) from t1;
 count  | count  
--------+--------
 105300 | 100000
(1 row)
```

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize is this still something you can reproduce on the latest main? I just tried `bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --other-tag latest --scenario MySqlStreaming` on main since your branch merged, and the first measurement seemed to work:

Let me check! (By the way, I got counts in different runs.)

**Comment by nrainer-materialize** [comment]
> Let me check! (By the way, I got different counts in different runs.)
> 
> Trying it out [here in CI](https://github.com/MaterializeInc/materialize/pull/25180) and I will also try it locally.

### CI
In CI, the triggered job is [feature benchmark #6.](https://buildkite.com/materialize/nightlies/builds/6414#018da190-2aaa-435a-bb86-58c24d954c9a) I will trigger a few retries. The first two builds passed. The failure seems to occur less frequently than in former build [builds/6375.](https://buildkite.com/materialize/nightlies/builds/6375)

### Locally
I managed to get it to fail locally. It was running for longer (first measurement completed successfully) until the error occurred (compared to earlier runs when working on the PR):
```
> SELECT count(*) FROM t1
/* B */
rows didn't match; sleeping to see if dataflow catches up 10ms

==> mzcompose: test case workflow-default failed: <class 'materialize.ui.CommandFailureCausedUIError'>: running docker compose failed (exit status 1)
mzcompose: error: at least one test case failed
nrainer@mz-nrainer mz-repo-2 % 
```

```
materialize=> SELECT count(*), count(distinct pk) FROM t1;
 count  | count  
--------+--------
 113500 | 100000
(1 row)
```

**Comment by petrosagg** [comment]
My 2c is that we should grab the commit that this reproduces on and figure out what's going on. Afaik we didn't make any substantial changes to the source and the issue not reproducing on `main` might just mean we're unlucky or something else hides the real issue.

**Comment by nrainer-materialize** [comment]
I will try to find that commit.

**Comment by rjobanp** [comment]
I've been trying to repro this locally with no success yet -- both on the commit `8f4d5b19763a3dc2e6a8415e91385abbd55eaf73` from your scenario PR and from the `7a402c18b6e1da752e2f46cfaf23e5730b465d9c` commit in the original build failures.

Tried running both `bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --other-tag latest --scenario MySqlStreaming` and also directly writing a testdrive scenario in the mysql-cdc tests (even with 1M rows):

```
def workflow_streaming(c: Composition, parser: WorkflowArgumentParser) -> None:
    c.up("materialized", "mysql")
    c.up("testdrive", persistent=True)

    n = 1000000

    insertions = "\n".join(
        [
            dedent(
                f"""
                SET @i:=0;
                INSERT INTO t1 (f2) SELECT @i:=@i+1 FROM mysql.time_zone t1, mysql.time_zone t2 LIMIT {round(n/1000)};
                COMMIT;
                """
            )
            for i in range(0, 1000)
        ]
    )

    c.testdrive(
        f"""
> SELECT 1;
1

$ postgres-execute connection=postgres://mz_system:materialize@${{testdrive.materialize-internal-sql-addr}}
ALTER SYSTEM SET enable_mysql_source = true

$ mysql-connect name=mysql url=mysql://root@mysql password={MySql.DEFAULT_ROOT_PASSWORD}

$ mysql-execute name=mysql
DROP DATABASE IF EXISTS public;
CREATE DATABASE public;
USE public;
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (pk SERIAL PRIMARY KEY, f2 BIGINT);

> CREATE SECRET IF NOT EXISTS mysqlpass AS '{MySql.DEFAULT_ROOT_PASSWORD}'
> CREATE CONNECTION IF NOT EXISTS mysql_conn TO MYSQL (
    HOST mysql,
    USER root,
    PASSWORD SECRET mysqlpass
  )

> CREATE CLUSTER source_cluster SIZE '1', REPLICATION FACTOR 1;

> CREATE SOURCE s1
    IN CLUSTER source_cluster
    FROM MYSQL CONNECTION mysql_conn
    FOR TABLES (public.t1);

> SELECT count(*) FROM t1
0

$ mysql-execute name=mysql
USE public;
{insertions}

> SELECT count(*) FROM t1
{n}
"""
    )
```

**Comment by nrainer-materialize** [comment]
> I've been trying to repro this locally with no success yet -- both on the commit `8f4d5b19763a3dc2e6a8415e91385abbd55eaf73` from your scenario PR and from the `7a402c18b6e1da752e2f46cfaf23e5730b465d9c` commit in the original build failures.
> Tried running both bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --other-tag latest --scenario MySqlStreaming and also directly writing a testdrive scenario in the mysql-cdc tests (even with 1M rows):

The feature benchmark is running all scenarios for two different versions of mz. Once for your checked out commit and once for the comparison version (baseline). Therefore, it is absolutely possible that you cannot reproduce the failure with the command above because `latest` (= `other-tag`) has changed. Let me try to find the problematic commit today by explicitly specifying the other target.

**Comment by nrainer-materialize** [comment]
The issue was fixed with https://github.com/MaterializeInc/materialize/pull/25054.

The problem is present in 32296a270b05b3761b65af184b0c3e887d21f9db but no longer in 156b015eb4d386b505d33c797278f08c2568f0ac.

**Comment by petrosagg** [comment]
@nrainer-materialize does the feature benchmark restart mz or clusterd? Because #25054 should only affect correctness in scenarios where restarts happen

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize does the feature benchmark restart mz or clusterd? Because #25054 should only affect correctness in scenarios where restarts happen

No, it does not. I also double-checked the [log of the failing execution (feature benchmark #6).](https://buildkite.com/materialize/nightlies/builds/6375#018d8dfc-c66c-48c1-a14d-6ad1bdcf4a3c)

Inserts happen with
```
SET @i:=0;
INSERT INTO t1 (f2) SELECT @i:=@i+1 FROM mysql.time_zone t1, mysql.time_zone t2 LIMIT 100;
COMMIT;
```

(100 rows per insert statement and transaction) and the select query happens directly after the last insert:
```
> SELECT count(*) FROM t1
 /* B */
100000

Discarding first measurement.
0    1104.355(memory)
1      10.151(wallclock)
1    1375.198(memory)
2      10.186(wallclock)
2    1592.636(memory)
^^^ +++
5011:1: error: non-matching rows: expected:
[["100000"]]
got:
[["106700"]]
Poor diff:
- 100000
+ 106700
```

**Comment by nrainer-materialize** [comment]
Oh... we got it again... on `main`: https://buildkite.com/materialize/nightlies/builds/6429#018da79c-1476-4cd9-928a-a4971a8b29ce

**Comment by petrosagg** [comment]
> Oh... we got it again... on main:

Unfortunate but also consistent with https://github.com/MaterializeInc/materialize/pull/25054 not possibly having an effect here

**Comment by rjobanp** [comment]
Yes sorry -- I was referring to the latest build/6429 that happened on main this morning!

**Comment by rjobanp** [comment]
Okay I was able to replicate it with this diff on top of today's `main` at `ee4b69e6bb0aefc98dc04b29f2448814fd0dc040`. The diff is basically the same as a revert of https://github.com/MaterializeInc/materialize/pull/25054.

```
5021:1: error: non-matching rows: expected:
[["100000"]]
got:
[["101600"]]
Poor diff:
- 100000
+ 101600
```

```diff
diff --git a/src/storage/src/source/mysql/replication.rs b/src/storage/src/source/mysql/replication.rs
index 093af1866d..61842449ca 100644
--- a/src/storage/src/source/mysql/replication.rs
+++ b/src/storage/src/source/mysql/replication.rs
@@ -192,26 +192,7 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
             // Calculate the lowest frontier across all subsources, which represents the point which
             // we should start replication from.
             let resume_upper =
-                match Antichain::from_iter(subsource_resume_uppers.into_values().flatten()) {
-                    upper if upper == Antichain::from_elem(GtidPartition::minimum()) => {
-                        // If any subsource is at the minimum frontier then we are either starting
-                        // from scratch or at least one table has to complete its initial snapshot.
-                        //
-                        // In either case, we need to ensure that the resumption point is
-                        // deterministic and consistent across all tables, so that a user can't
-                        // observe an MZ timestamp that may contain different snapshots for
-                        // different tables.
-                        // We've chosen the frontier beyond the GTID Set recorded
-                        // during purification as this consistent resume point.
-                        //
-                        // Tables may still be snapshot at a future GTID Set but as long as we can
-                        // resume the replication stream from this consistent point all updates
-                        // will be rewound appropriately, such that the effective snapshot point
-                        // is consistent across all tables.
-                        gtid_set_frontier(&connection.details.initial_gtid_set)?
-                    }
-                    upper => upper,
-                };
+                Antichain::from_iter(subsource_resume_uppers.into_values().flatten());

             // Validate that we can actually resume from this upper.
             if !PartialOrder::less_equal(&binlog_frontier, &resume_upper) {
@@ -255,9 +236,9 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                             .await);
                         };
                         // If the snapshot point is the same as the resume point then we don't need to rewind
-                        if resume_upper != req.snapshot_upper {
-                            rewinds.insert(req.table.clone(), (caps.clone(), req));
-                        }
+                        // if resume_upper != req.snapshot_upper {
+                        rewinds.insert(req.table.clone(), (caps.clone(), req));
+                        // }
                     }
                 }
             }
@@ -338,10 +319,7 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                             upper_cap_set.downgrade(&*new_upper);

                             rewinds.retain(|_, (_, req)| {
-                                // We need to retain the rewind requests whose snapshot_upper has
-                                // at least one timestamp such that new_upper is less than that
-                                // timestamp
-                                req.snapshot_upper.iter().any(|ts| new_upper.less_than(ts))
+                                !PartialOrder::less_equal(&req.snapshot_upper, &new_upper)
                             });

                             trace!(%id, "timely-{worker_id} pending rewinds after \
@@ -471,10 +449,7 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                         upper_cap_set.downgrade(&*new_upper);

                         rewinds.retain(|_, (_, req)| {
-                            // We need to retain the rewind requests whose snapshot_upper has
-                            // at least one timestamp such that new_upper is less than that
-                            // timestamp
-                            req.snapshot_upper.iter().any(|ts| new_upper.less_than(ts))
+                            !PartialOrder::less_equal(&req.snapshot_upper, &new_upper)
                         });
                         trace!(%id, "timely-{worker_id} pending rewinds \
                                      after filtering: {rewinds:?}");
```

**Comment by nrainer-materialize** [comment]
Just to confirm: This does mean that https://github.com/MaterializeInc/materialize/pull/25054 fixed the issue?

**Comment by rjobanp** [comment]
I'm 90% sure that PR fixed the issue.

I've been able to isolate the diff that reproduces the issue to this change
```
diff --git a/src/storage/src/source/mysql/replication.rs b/src/storage/src/source/mysql/replication.rs
index 093af1866d..61842449ca 100644
--- a/src/storage/src/source/mysql/replication.rs
+++ b/src/storage/src/source/mysql/replication.rs
@@ -192,26 +192,7 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
             // Calculate the lowest frontier across all subsources, which represents the point which
             // we should start replication from.
             let resume_upper =
-                match Antichain::from_iter(subsource_resume_uppers.into_values().flatten()) {
-                    upper if upper == Antichain::from_elem(GtidPartition::minimum()) => {
-                        // If any subsource is at the minimum frontier then we are either starting
-                        // from scratch or at least one table has to complete its initial snapshot.
-                        //
-                        // In either case, we need to ensure that the resumption point is
-                        // deterministic and consistent across all tables, so that a user can't
-                        // observe an MZ timestamp that may contain different snapshots for
-                        // different tables.
-                        // We've chosen the frontier beyond the GTID Set recorded
-                        // during purification as this consistent resume point.
-                        //
-                        // Tables may still be snapshot at a future GTID Set but as long as we can
-                        // resume the replication stream from this consistent point all updates
-                        // will be rewound appropriately, such that the effective snapshot point
-                        // is consistent across all tables.
-                        gtid_set_frontier(&connection.details.initial_gtid_set)?
-                    }
-                    upper => upper,
-                };
+                Antichain::from_iter(subsource_resume_uppers.into_values().flatten());
```

which causes the `resume_upper` to be set to an `Antichain` of `[GtidPartition::minimum()]`, which then causes us to pass an empty vector to the `BinlogStreamRequest` `with_gtid_set` argument:
https://github.com/MaterializeInc/materialize/blob/668b1cabb8e472510c62aef9352446062c2f5ad7/src/storage/src/source/mysql/replication.rs#L597-L602

If you track that into the `mysql-common` crate, using an empty vector will cause it to remove the `BINLOG_THROUGH_GTID` flag on the `ComBinlogDumpGtid` message it sends to request the replication stream:
https://github.com/blackbeam/rust_mysql_common/blob/48ed958051d9a3df1ed1bec3ca0f893b2a18dcc4/src/packets/mod.rs#L3278-L3287

In the mysql-server source, this is where it handles the `ComBinlogDumpGtid` message:
https://github.com/mysql/mysql-server/blob/824e2b4064053f7daf17d7f3f84b7a3ed92e5fb4/sql/rpl_source.cc#L985

which parses the flags and passes them to the `Binlog_sender` class: https://github.com/mysql/mysql-server/blob/49ef33f7edadef3ae04665e73d1babd40179a4f1/sql/rpl_binlog_sender.cc#L233-L235

but I can't find anything in the mysql-server source that actually looks for that specific flag (so I'm not convinced it actually does anything!).

I'm still wondering if the flag is causing some sort of replication stream behavior difference, or if the fact that we're not starting replication at the beginning of the binlog (and therefore avoiding the need to rewind updates) is what fixed the issue. If so, then we need to make sure our rewind logic is still correct.

**Comment by rjobanp** [comment]
Okay now I think that PR doesn't fix the issue, it just makes it much less likely to occur since we start replication from the same point that the snapshot occurred at, rather than the beginning of the binlog each time.

I was able to reproduce on the latest main by forcing the `resume_upper` to start at a point 100 GTIDs before the snapshot point, so I think the issue is present somewhere in our rewind logic.

**Comment by rjobanp** [comment]
Shoot, closing was a mistake! I tagged the wrong issue in this PR https://github.com/MaterializeInc/materialize/pull/25245

**Comment by rjobanp** [comment]
After lots of hypothesis testing and validation, we've come to the following conclusions:

On reading `@gtid_executed`:
- We can't trust that `LOCK TABLES {table} READ, ...` and then immediately reading `@gtid_executed` will give us an accurate gtid_executed value since gtid_executed is updated non-atomically. However using `FLUSH TABLES {table}, .. WITH READ LOCK` fixes this by allowing us to start a transaction with a consistent view of the table and read the gtid_executed value in a consistent way. This is documented in this commit to mysql-server: https://github.com/mysql/mysql-server/commit/022e73ba6976b984658a1c2652178cd4b81aec28

On consistent reads:
- We can't trust that a `START TRANSACTION .. WITH CONSISTENT SNAPSHOT` done while the tables are locked will actually give us a consistent read once the tables are unlocked. We observed in testing that the `SELECT *` in that transaction (after tables got unlocked in the separate connection) would give us rows from more transactions than were originally observed when the table was locked.
- We instead tried to do a `SELECT * FROM {table} LIMIT 1` (before the `UNLOCK TABLES` happened in the separate connection) to trigger a consistent read, and this resulted in the same error.
- However, we found that a `SELECT count(*) FROM {table}` in the transaction, followed by the `UNLOCK TABLES` in the separate connection would give us the correct result that corresponded with the transactions in the `@gtid_executed` set after the lock was taken.
- It's possible that we could simplify our code to use one connection rather than 2, and have the single connection start the repeatable read transaction, flush/lock the tables, observe gtid_executed, run the `SELECT count(*) ..`, unlock the tables, and then read the full snapshot. However the MySQL docs are a little ambiguous on how the `UNLOCK TABLES` interacts with the transaction and whether it might implicitly commit the connection's transaction, so for now we are going to stick with the 2-connection approach.

I'll have a PR up momentarily that implements these two changes to solve this issue.

**Comment by rjobanp** [comment]
Okay, all of my conclusions above were incorrect and the actual bug was a simple misuse of the `mysql-async` `Conn.start_transaction` api:
![image](https://github.com/MaterializeInc/materialize/assets/3818834/78b79bdd-d908-44f8-aab6-64d3e1577428)

We ended up seeing all of those consistency issues above because we weren't actually inside a transaction when performing the snapshot! Thanks to @benesch for noticing this -- fix is in #25312
